File: Design-Patterns-Using-Go/Flyweight/FlyweightPattern.go
package main

import (
	"fmt"
)

// Flyweight interface
type Flyweight interface {
	Operation(extrinsicState interface{})
}

// ConcreteFlyweight type
type ConcreteFlyweight struct {
	intrinsicState string
}

func (cf *ConcreteFlyweight) Operation(extrinsicState interface{}) {
	fmt.Println("Operation inside concrete flyweight")
}

// FlyweightFactory type
type FlyweightFactory struct {
	flyweights map[string]Flyweight
}

func NewFlyweightFactory() *FlyweightFactory {
	return &FlyweightFactory{
		flyweights: make(map[string]Flyweight),
	}
}

func (ff *FlyweightFactory) GetFlyweight(key string) Flyweight {
	if flyweight, exists := ff.flyweights[key]; exists {
		return flyweight
	}

	flyweight := &ConcreteFlyweight{intrinsicState: key}
	ff.flyweights[key] = flyweight
	return flyweight
}

func (ff *FlyweightFactory) GetCount() int {
	return len(ff.flyweights)
}

// Client Code
func main() {
	factory := NewFlyweightFactory()
	flyweight1 := factory.GetFlyweight("key")
	flyweight2 := factory.GetFlyweight("key")

	flyweight1.Operation(nil)
	fmt.Println(flyweight1, flyweight2)
	fmt.Println("Object count:", factory.GetCount())
}

/*
Operation inside concrete flyweight
&{key} &{key}
Object count: 1
*/

File: Design-Patterns-Using-Cpp/Flyweight/FlyweightPattern.cpp
#include <iostream>
#include <string>
#include <unordered_map>

// Flyweight abstract class
class Flyweight {
protected:
    std::string intrinsicState;

public:
    Flyweight(const std::string& intrinsicState) : intrinsicState(intrinsicState) {}

    virtual void operation(const void* extrinsicState) = 0;
};

// Concrete Flyweight class
class ConcreteFlyweight : public Flyweight {
public:
    ConcreteFlyweight(const std::string& intrinsicState) : Flyweight(intrinsicState) {}

    void operation(const void* extrinsicState) override {
        std::cout << "Operation inside concrete flyweight\n";
    }
};

// FlyweightFactory class
class FlyweightFactory {
private:
    std::unordered_map<std::string, Flyweight*> flyweights;

public:
    Flyweight* getFlyweight(const std::string& key) {
        if (flyweights.find(key) == flyweights.end()) {
            flyweights[key] = new ConcreteFlyweight(key);
        }
        return flyweights[key];
    }

    int getCount() const {
        return flyweights.size();
    }

    ~FlyweightFactory() {
        for (auto& pair : flyweights) {
            delete pair.second;
        }
    }
};

// Client code
int main() {
    FlyweightFactory factory;
    Flyweight* flyweight1 = factory.getFlyweight("key");
    Flyweight* flyweight2 = factory.getFlyweight("key");
    flyweight1->operation(nullptr);
    std::cout << flyweight1 << " " << flyweight2 << std::endl;
    std::cout << "Object count: " << factory.getCount() << std::endl;

    return 0;
}


/*
Operation inside concrete flyweight
0x55990ff3beb0 0x55990ff3beb0
Object count: 1
*/